### PR TITLE
Autologging: Add context to MLflow warnings raised by patch implementations

### PR DESCRIPTION
**Pull request by dbczumar**
## What changes are proposed in this pull request?

MLflow routines called by autologging patch code may issue warnings via the `warnings.warn` API. In many cases, the user cannot remediate the cause of these warnings because they result from the autologging patch implementation, rather than a user-facing API call.

Accordingly, this PR introduces a context manager that is designed to augment MLflow warnings issued during
autologging patch code execution, explaining the warnings were raised as a result of MLflow's autologging implementation. MLflow warnings are also redirected from `sys.stderr` to an MLflow logger with level WARNING. Warnings issued by code outside of MLflow are not modified. When the context manager exits, the original output behavior for MLflow warnings
is restored.

On Databricks, this helps reroute potentially confusing warning messages that may arise during autologging; e.g., warnings about integer columns during schema inference.

**Example - run the following workflow:**

```
import mlflow
mlflow.autolog()

from sklearn import datasets
iris = datasets.load_iris()

from sklearn.linear_model import LogisticRegression
from scipy.stats import uniform

logistic = LogisticRegression(solver="saga", tol=1e-2, max_iter=200, random_state=0)
train_out = logistic.fit(iris.data, iris.target)
```

**Before**:
```
2020/12/21 01:22:43 INFO mlflow.tracking.fluent: Autologging successfully enabled for sklearn.
/Users/czumar/mlflow/mlflow/models/signature.py:124: UserWarning: Hint: Inferred schema contains integer column(s).
 Integer columns in Python cannot represent missing values. If your input data contains missing values at inference time, it
 will be encoded as floats and will cause a schema enforcement error. The best way to avoid this problem is to infer the
 model schema based on a realistic data sample (training dataset) that includes missing values. Alternatively, you can
 declare integer columns as doubles (float64) whenever these columns may have missing values. See `Handling Integers
 With Missing Values <https://www.mlflow.org/docs/latest/models.html#handling-integers-with-missing-values>`_ for more
 details.
 outputs = _infer_schema(model_output) if model_output is not None else None
```

**After**:
```
2020/12/21 01:23:12 INFO mlflow.tracking.fluent: Autologging successfully enabled for sklearn.
2020/12/21 01:23:13 WARNING mlflow.utils.autologging_utils: MLflow issued a warning during sklearn autologging:
 "/Users/czumar/mlflow/mlflow/models/signature.py:124: UserWarning: Hint: Inferred schema contains integer column(s).
 Integer columns in Python cannot represent missing values. If your input data contains missing values at inference time, it
 will be encoded as floats and will cause a schema enforcement error. The best way to avoid this problem is to infer the
 model schema based on a realistic data sample (training dataset) that includes missing values. Alternatively, you can
 declare integer columns as doubles (float64) whenever these columns may have missing values. See `Handling Integers
 With Missing Values <https://www.mlflow.org/docs/latest/models.html#handling-integers-with-missing-values>`_ for more
 details."
```

## How is this patch tested?

New unit tests for warning behavior, existing unit + integration tests for correctness

## Release Notes

Improve clarity of warning messages raised during autologging

### Is this a user-facing change?

- [ ] No. You can skip the rest of this section.
- [X] Yes. Give a description of this change to be included in the release notes for MLflow users.

### What component(s), interfaces, languages, and integrations does this PR affect?
Components 
- [ ] `area/artifacts`: Artifact stores and artifact logging
- [ ] `area/build`: Build and test infrastructure for MLflow
- [ ] `area/docs`: MLflow documentation pages
- [ ] `area/examples`: Example code
- [ ] `area/model-registry`: Model Registry service, APIs, and the fluent client calls for Model Registry
- [ ] `area/models`: MLmodel format, model serialization/deserialization, flavors
- [ ] `area/projects`: MLproject format, project running backends
- [ ] `area/scoring`: Local serving, model deployment tools, spark UDFs
- [ ] `area/server-infra`: MLflow server, JavaScript dev server
- [X] `area/tracking`: Tracking Service, tracking client APIs, autologging

Interface 
- [ ] `area/uiux`: Front-end, user experience, JavaScript, plotting
- [ ] `area/docker`: Docker use across MLflow's components, such as MLflow Projects and MLflow Models
- [ ] `area/sqlalchemy`: Use of SQLAlchemy in the Tracking Service or Model Registry
- [ ] `area/windows`: Windows support

Language 
- [ ] `language/r`: R APIs and clients
- [ ] `language/java`: Java APIs and clients
- [ ] `language/new`: Proposals for new client languages

Integrations
- [ ] `integrations/azure`: Azure and Azure ML integrations
- [ ] `integrations/sagemaker`: SageMaker integrations
- [ ] `integrations/databricks`: Databricks integrations

### How should the PR be classified in the release notes? Choose one:

- [ ] `rn/breaking-change` - The PR will be mentioned in the "Breaking Changes" section
- [X] `rn/none` - No description will be included. The PR will be mentioned only by the PR number in the "Small Bugfixes and Documentation Updates" section
- [ ] `rn/feature` - A new user-facing feature worth mentioning in the release notes
- [ ] `rn/bug-fix` - A user-facing bug fix worth mentioning in the release notes
- [ ] `rn/documentation` - A user-facing documentation change worth mentioning in the release notes

### CODE REVIEW

File: mlflow/utils/autologging_utils.py [hunk 1]
```diff
@@ -1041,6 +1043,52 @@ def call_original(*og_args, **og_kwargs):
...
+    :param autologging_integration: The name of the active autologging integration for which
+                                    MLflow warnings are to be augmented during patch code
+                                    execution.
+    """
+    original_showwarning = warnings.showwarning
+
+    def autologging_showwarning(message, category, filename, lineno, *args, **kwargs):
+        mlflow_root_path = Path(os.path.dirname(mlflow.__file__)).resolve()
+        warning_source_path = Path(filename).resolve()
+        if mlflow_root_path in warning_source_path.parents:
```
Comment (apurva-koti) [inline mlflow/utils/autologging_utils.py:1069]: Is this a way of checking if the file from where the warning originated is a descendant of the mlflow directory?

Comment (dbczumar) [reply mlflow/utils/autologging_utils.py:1069]: Yessir!

File: mlflow/utils/autologging_utils.py [hunk 2]
```diff
@@ -1041,6 +1043,52 @@ def call_original(*og_args, **og_kwargs):
...
+            _logger.warning(
+                "MLflow issued a warning during %s autologging:" ' "%s:%d: %s: %s"',
+                autologging_integration,
+                filename,
+                lineno,
+                category.__name__,
+                message,
+            )
+        else:
+            original_showwarning(message, category, filename, *args, **kwargs)
```
Comment (dbczumar) [inline mlflow/utils/autologging_utils.py:1079]: Whoops! Missing lineno here...

File: mlflow/utils/autologging_utils.py (hunk 1 above)
Comment (dbczumar) [reply mlflow/utils/autologging_utils.py:1069]: Will add an explanatory comment!

File: tests/autologging/test_autologging_safety_unit.py [hunk 3]
```diff
@@ -734,6 +735,66 @@ def patch_impl(original, *args, **kwargs):
...
+
+    with pytest.warns(UserWarning) as user_warnings_outside_patch:
+        warnings.warn_explicit(**mlflow_warning_kwargs)
+        warnings.warn_explicit(**external_warning_kwargs)
+
+    # Verify that MLflow warnings and external warnings are emitted as normal outside
+    # of autologging patch execution
+    assert set([warning.message.args[0] for warning in user_warnings_outside_patch]) == set(
+        ["Mock MLflow warning", "Mock external warning"]
+    )
```
Comment (apurva-koti) [inline tests/autologging/test_autologging_safety_unit.py:795]: Nice :)

File: mlflow/utils/autologging_utils.py (hunk 2 above)
Comment (apurva-koti) [reply mlflow/utils/autologging_utils.py:1079]: Whoops! Missed that myself 😆

File: mlflow/utils/autologging_utils.py (hunk 1 above)
Comment (apurva-koti) [reply mlflow/utils/autologging_utils.py:1069]: Just curious, is there any other way of doing this? Never seen this approach before. (Don't really have an issue with it myself though)

**Review by dbczumar — COMMENTED** [review]
@apurva-koti Addressed your comment, fixed the `lineno` issue, and improved test coverage for various warning attributes (lineno, file name, message)

**Review by apurva-koti — APPROVED** [review]
Nice!! 🍔

**Pull request merged**